lib/puppet/network/http/response.rb

```diff
@@ -5,7 +5,23 @@ def initialize(handler, response)
   end
 
   def respond_with(code, type, body)
-    @handler.set_content_type(@response, type)
+    format = Puppet::Network::FormatHandler.format_for(type)
+    mime = format.mime
+    charset = format.charset
+
+    if charset && body.is_a?(String)
+      # REMIND: not all charsets are valid ruby encodings, e.g. ISO-2022-KR
+      encoding = Encoding.find(charset)
+
+      if body.encoding != encoding
+        # REMIND this can raise if body contains invalid UTF-8
+        body.encode!(encoding)
```

Comment:
Add a test that `respond_with` raises if given data that can't be transcoded to UTF-8